File: services/api/app/services/risk_engine.py
from __future__ import annotations

import re
from dataclasses import dataclass

from app.schemas.jobs import PreviewAnalysisResponse, RiskClassification, RiskFactor

SENTENCE_SPLIT_PATTERN = re.compile(r"(?<=[.!?])\s+")


@dataclass(frozen=True)
class RiskRule:
    key: str
    label: str
    weight: int
    explanation: str
    patterns: tuple[re.Pattern[str], ...]


def _compile(*expressions: str) -> tuple[re.Pattern[str], ...]:
    return tuple(re.compile(expression, re.IGNORECASE) for expression in expressions)


RISK_RULES: tuple[RiskRule, ...] = (
    RiskRule(
        key="auto_renewal",
        label="Auto-renewal",
        weight=18,
        explanation="The contract may renew unless the user gives advance notice.",
        patterns=_compile(r"automatic(?:ally)? renew", r"renews? automatically"),
    ),
    RiskRule(
        key="cancellation_penalty",
        label="Cancellation penalty",
        weight=16,
        explanation="Leaving the agreement early may trigger extra fees or penalties.",
        patterns=_compile(r"penalt(?:y|ies) for early cancellation", r"termination fee"),
    ),
    RiskRule(
        key="hidden_fees",
        label="Hidden fees",
        weight=14,
        explanation="The contract references additional costs that may not be capped or clearly itemized.",
        patterns=_compile(r"additional fees?", r"administration fee", r"billed separately"),
    ),
    RiskRule(
        key="non_compete",
        label="Non-compete clause",
        weight=20,
        explanation="The agreement may restrict work or business activity after the relationship ends.",
        patterns=_compile(r"non-?compete", r"shall not engage in any competing business"),
    ),
    RiskRule(
        key="lock_in",
        label="Long-term lock-in",
        weight=15,
        explanation="The user may be tied into the contract for a long minimum period.",
        patterns=_compile(r"minimum term of \w+", r"locked in for", r"fixed term of \w+"),
    ),
    RiskRule(
        key="liability_imbalance",
        label="Liability imbalance",
        weight=17,
        explanation="Responsibility or damages appear to fall unevenly on one party.",
        patterns=_compile(r"indemnif(?:y|ication)", r"not liable for any loss", r"sole liability"),
    ),
    RiskRule(
        key="arbitration",
        label="Arbitration clause",
        weight=10,
        explanation="Disputes may be forced into arbitration instead of court.",
        patterns=_compile(r"binding arbitration", r"resolved by arbitration"),
    ),
)


def _extract_sentences(text: str) -> list[str]:
    normalized = " ".join(text.split())
    parts = SENTENCE_SPLIT_PATTERN.split(normalized)
    return [part.strip() for part in parts if part.strip()]


def _pick_matching_sentences(sentences: list[str], keywords: tuple[str, ...], limit: int = 2) -> list[str]:
    matches: list[str] = []
    for sentence in sentences:
        lowered = sentence.lower()
        if any(keyword in lowered for keyword in keywords):
            matches.append(sentence)
        if len(matches) >= limit:
            break
    return matches


def classify_risk(score: int) -> RiskClassification:
    if score >= 60:
        return RiskClassification.high
    if score >= 25:
        return RiskClassification.medium
    return RiskClassification.low


def _find_match(rule: RiskRule, text: str) -> re.Match[str] | None:
    for pattern in rule.patterns:
        match = pattern.search(text)
        if match:
            return match
    return None


def analyze_text(text: str) -> tuple[list[RiskFactor], int, RiskClassification]:
    normalized = " ".join(text.split())
    factors: list[RiskFactor] = []

    for rule in RISK_RULES:
        match = _find_match(rule, normalized)
        if not match:
            continue

        start = max(match.start() - 40, 0)
        end = min(match.end() + 70, len(normalized))
        evidence = normalized[start:end].strip()
        factors.append(
            RiskFactor(
                key=rule.key,
                label=rule.label,
                weight=rule.weight,
                evidence=evidence,
                explanation=rule.explanation,
            )
        )

    score = min(sum(factor.weight for factor in factors), 100)
    return factors, score, classify_risk(score)


def detect_contract_type(text: str) -> str:
    normalized = text.lower()
    contract_signals = (
        ("lease agreement", ("lease", "landlord", "tenant", "rent", "premises")),
        ("employment contract", ("employee", "employer", "salary", "non-compete", "termination")),
        ("service agreement", ("services", "service provider", "scope of work", "deliverables")),
        ("loan agreement", ("loan", "interest", "repayment", "borrower", "lender")),
        ("subscription agreement", ("subscription", "renew", "monthly plan", "membership")),
    )
    for label, keywords in contract_signals:
        if sum(keyword in normalized for keyword in keywords) >= 2:
            return label
    return "general contract"


def _build_recommended_actions(factors: list[RiskFactor], risk_level: RiskClassification) -> list[str]:
    actions: list[str] = []
    factor_keys = {factor.key for factor in factors}

    if "auto_renewal" in factor_keys:
        actions.append("Ask for the renewal notice period and whether renewal can be disabled before signing.")
    if "cancellation_penalty" in factor_keys or "lock_in" in factor_keys:
        actions.append("Clarify the exact cancellation cost, notice requirement, and minimum commitment period.")
    if "hidden_fees" in factor_keys:
        actions.append("Request a written fee schedule that lists every recurring and one-off charge.")
    if "liability_imbalance" in factor_keys:
        actions.append("Check whether liability and indemnity obligations can be made more balanced.")
    if risk_level == RiskClassification.high:
        actions.append("Consider pausing signature and escalating the contract for legal review.")

    if not actions:
        actions.append("Confirm the practical payment, renewal, and termination terms in writing before signing.")

    return actions[:3]


import instructor
from anthropic import Anthropic
from app.core.config import settings

def _build_claude_analysis(text: str, fallback_factors: list[RiskFactor]) -> PreviewAnalysisResponse:
    if not settings.CLAUDE_API_KEY or settings.CLAUDE_API_KEY == "sk_test_mock":
        # Fallback to deterministic logic if no live key is mapped.
        return _build_preview_analysis_deterministic(text, fallback_factors)

    client = instructor.from_anthropic(Anthropic(api_key=settings.CLAUDE_API_KEY))
    
    prompt = f"""
    You are an expert legal contract analyzer. Your job is to extract clauses and risks from the following contract.
    We have already detected some baseline risk factors using deterministic rules. Please merge your findings with ours.
    Be objective, accurate, and map EXACTLY to the requested JSON schema.
    
    Contract text:
    {text}
    """
    
    resp = client.messages.create(
        model="claude-3-haiku-20240307",
        max_tokens=2048,
        messages=[{"role": "user", "content": prompt}],
        response_model=PreviewAnalysisResponse
    )
    return resp

def _build_preview_analysis_deterministic(text: str, factors: list[RiskFactor]) -> PreviewAnalysisResponse:
    score = min(sum(factor.weight for factor in factors), 100)
    risk_level = classify_risk(score)
    sentences = _extract_sentences(text)
    contract_type = detect_contract_type(text)

    if factors:
        factor_list = ", ".join(factor.label.lower() for factor in factors[:3])
        summary = (
            f"The preview found signals related to {factor_list}. "
            "A full AI review should confirm whether those clauses are enforceable and how they interact."
        )
    else:
        summary = (
            "The preview did not match the launch-rule risk patterns, but the contract should still go through full AI analysis."
        )

    key_points = sentences[:3] if sentences else ["Normalized contract text will appear here after extraction."]
    financial_obligations = _pick_matching_sentences(sentences, keywords=("fee", "cost", "payment", "rent", "billed", "price"))
    duration_terms = _pick_matching_sentences(sentences, keywords=("term", "month", "year", "renew", "duration"))
    cancellation_terms = _pick_matching_sentences(sentences, keywords=("cancel", "cancellation", "terminate", "notice", "exit"))

    red_flags = [f"{factor.label}: {factor.explanation} Evidence: {factor.evidence}" for factor in factors]

    return PreviewAnalysisResponse(
        contract_type=contract_type,
        summary=summary,
        key_points=key_points,
        red_flags=red_flags,
        risk_score=score,
        risk_level=risk_level,
        factors=factors,
        financial_obligations=financial_obligations,
        duration_terms=duration_terms,
        cancellation_terms=cancellation_terms,
        recommended_actions=_build_recommended_actions(factors, risk_level),
    )

def build_preview_analysis(text: str) -> PreviewAnalysisResponse:
    # 1. Run deterministic regex
    factors, _, _ = analyze_text(text)
    
    # 2. Hand over to Claude for holistic extraction and merge
    return _build_claude_analysis(text, factors)

File: apps/web/app/lib/risk-engine.ts
import type {
  PreviewAnalysisResponse,
  RiskClassification,
  RiskFactor,
} from "./types";

const riskRules = [
  [
    "auto_renewal",
    "Auto-renewal",
    18,
    "The contract may renew unless the user gives advance notice.",
    [/automatic(?:ally)? renew/i, /renews? automatically/i, /shall automatically renew/i],
  ],
  [
    "cancellation_penalty",
    "Cancellation penalty",
    16,
    "Leaving the agreement early may trigger extra fees or penalties.",
    [/penalt(?:y|ies) for early cancellation/i, /termination fee/i, /early termination/i],
  ],
  [
    "hidden_fees",
    "Hidden fees",
    14,
    "The contract references additional costs that may not be capped or clearly itemized.",
    [/additional fees?/i, /administration fee/i, /billed separately/i, /service charge/i],
  ],
  [
    "non_compete",
    "Non-compete clause",
    20,
    "The agreement may restrict work or business activity after the relationship ends.",
    [/non-?compete/i, /shall not engage in any competing business/i, /restraint of trade/i],
  ],
  [
    "lock_in",
    "Long-term lock-in",
    15,
    "The user may be tied into the contract for a long minimum period.",
    [/minimum term of \w+/i, /locked in for/i, /fixed term of \w+/i, /minimum commitment/i],
  ],
  [
    "liability_imbalance",
    "Liability imbalance",
    17,
    "Responsibility or damages appear to fall unevenly on one party.",
    [/indemnif(?:y|ication)/i, /not liable for any loss/i, /sole liability/i, /hold harmless/i],
  ],
  [
    "arbitration",
    "Arbitration clause",
    10,
    "Disputes may be forced into arbitration instead of court.",
    [/binding arbitration/i, /resolved by arbitration/i, /waive.*right.*jury/i],
  ],
] as const;

export function normalizeText(text: string) {
  return text.replace(/\s+/g, " ").trim();
}

export function splitSentences(text: string) {
  return normalizeText(text)
    .split(/(?<=[.!?])\s+/)
    .map((sentence) => sentence.trim())
    .filter(Boolean);
}

export function detectContractType(text: string) {
  const normalized = text.toLowerCase();
  const contracts = [
    ["lease agreement", ["lease", "landlord", "tenant", "rent", "premises"]],
    ["employment contract", ["employee", "employer", "salary", "non-compete", "termination"]],
    ["service agreement", ["services", "service provider", "scope of work", "deliverables"]],
    ["loan agreement", ["loan", "interest", "repayment", "borrower", "lender"]],
    ["subscription agreement", ["subscription", "renew", "monthly plan", "membership"]],
    ["non-disclosure agreement", ["confidential", "non-disclosure", "nda", "proprietary"]],
  ] as const;

  for (const [label, keywords] of contracts) {
    if (keywords.filter((keyword) => normalized.includes(keyword)).length >= 2) {
      return label;
    }
  }

  return "general contract";
}

export function classifyRisk(score: number): RiskClassification {
  if (score >= 60) return "high";
  if (score >= 25) return "medium";
  return "low";
}

function pickMatchingSentences(sentences: string[], keywords: string[]) {
  return sentences
    .filter((sentence) => keywords.some((keyword) => sentence.toLowerCase().includes(keyword)))
    .slice(0, 2);
}

function buildRecommendedActions(factors: RiskFactor[], riskLevel: RiskClassification) {
  const keys = new Set(factors.map((factor) => factor.key));
  const actions: string[] = [];

  if (keys.has("auto_renewal"))
    actions.push("Ask for the renewal notice period and whether renewal can be disabled before signing.");
  if (keys.has("cancellation_penalty") || keys.has("lock_in"))
    actions.push("Clarify the exact cancellation cost, notice requirement, and minimum commitment period.");
  if (keys.has("hidden_fees"))
    actions.push("Request a written fee schedule that lists every recurring and one-off charge.");
  if (keys.has("liability_imbalance"))
    actions.push("Check whether liability and indemnity obligations can be made more balanced.");
  if (keys.has("non_compete"))
    actions.push("Negotiate the non-compete scope, geography, and duration before accepting.");
  if (riskLevel === "high")
    actions.push("Consider pausing signature and escalating the contract for legal review.");
  if (actions.length === 0)
    actions.push("Confirm the practical payment, renewal, and termination terms in writing before signing.");

  return actions.slice(0, 4);
}

export function buildRuleAnalysis(text: string): PreviewAnalysisResponse {
  const normalized = normalizeText(text);
  const sentences = splitSentences(normalized);

  const factors: RiskFactor[] = riskRules.flatMap(([key, label, weight, explanation, patterns]) => {
    const match = patterns.map((pattern) => pattern.exec(normalized)).find(Boolean);
    if (!match || match.index === undefined) return [];

    return [
      {
        key,
        label,
        weight,
        evidence: normalized
          .slice(Math.max(match.index - 40, 0), Math.min(match.index + match[0].length + 70, normalized.length))
          .trim(),
        explanation,
      },
    ];
  });

  const riskScore = Math.min(
    factors.reduce((total, factor) => total + factor.weight, 0),
    100,
  );
  const riskLevel = classifyRisk(riskScore);

  return {
    contract_type: detectContractType(text),
    summary:
      factors.length > 0
        ? `The preview found signals related to ${factors
            .slice(0, 3)
            .map((factor) => factor.label.toLowerCase())
            .join(", ")}. A full review should confirm whether those clauses are enforceable and how they interact.`
        : "The preview did not match the launch-rule risk patterns, but the contract should still go through full AI analysis.",
    key_points:
      sentences.length > 0 ? sentences.slice(0, 3) : ["Normalized contract text will appear here after extraction."],
    red_flags: factors.map((factor) => `${factor.label}: ${factor.explanation} Evidence: ${factor.evidence}`),
    risk_score: riskScore,
    risk_level: riskLevel,
    factors,
    financial_obligations: pickMatchingSentences(sentences, [
      "fee",
      "fees",
      "cost",
      "costs",
      "payment",
      "rent",
      "billed",
      "price",
      "penalty",
      "salary",
    ]),
    duration_terms: pickMatchingSentences(sentences, [
      "term",
      "month",
      "months",
      "year",
      "years",
      "renew",
      "renewal",
      "duration",
    ]),
    cancellation_terms: pickMatchingSentences(sentences, [
      "cancel",
      "cancellation",
      "terminate",
      "termination",
      "notice",
      "exit",
    ]),
    recommended_actions: buildRecommendedActions(factors, riskLevel),
  };
}

export function buildRuleFollowUpAnswer(analysis: PreviewAnalysisResponse, question: string) {
  const normalized = question.toLowerCase();
  const firstAction = analysis.recommended_actions[0] ?? null;

  if (normalized.includes("cancel") || normalized.includes("termination") || normalized.includes("exit")) {
    const factor = analysis.factors.find((item) =>
      ["cancellation_penalty", "lock_in", "auto_renewal"].includes(item.key),
    );
    return {
      answer: factor
        ? `${factor.label}: ${factor.explanation}`
        : "The preview did not find a strong cancellation signal, so confirm notice periods and exit costs in the full contract.",
      suggested_next_step: firstAction,
    };
  }
  if (normalized.includes("renew")) {
    const factor = analysis.factors.find((item) => ["auto_renewal", "lock_in"].includes(item.key));
    return {
      answer: factor
        ? `${factor.label}: ${factor.explanation}`
        : "The preview did not flag renewal language, but the full review should still confirm whether the contract continues automatically.",
      suggested_next_step: firstAction,
    };
  }
  if (
    normalized.includes("fee") ||
    normalized.includes("cost") ||
    normalized.includes("pay") ||
    normalized.includes("money")
  ) {
    const factor = analysis.factors.find((item) =>
      ["hidden_fees", "cancellation_penalty"].includes(item.key),
    );
    const paymentLine = analysis.financial_obligations[0];
    return {
      answer: paymentLine
        ? `${factor ? `${factor.label}: ${factor.explanation} ` : ""}Key payment language: ${paymentLine}`
        : "The preview did not isolate a clear payment clause, so ask for a full fee breakdown before signing.",
      suggested_next_step: firstAction,
    };
  }
  if (normalized.includes("safe") || normalized.includes("sign") || normalized.includes("agree")) {
    if (analysis.risk_level === "high")
      return {
        answer: `This preview leans high risk, so signing without clarification would be hard to recommend. Focus first on: ${
          firstAction ?? "clarifying the flagged terms."
        }`,
        suggested_next_step: firstAction,
      };
    if (analysis.risk_level === "medium")
      return {
        answer: `This contract may still be workable, but only after clarifying the flagged terms. Start with: ${
          firstAction ?? "the main obligations."
        }`,
        suggested_next_step: firstAction,
      };
    return {
      answer:
        "The preview looks relatively low risk, but it is still worth confirming the practical obligations in writing before you sign.",
      suggested_next_step: firstAction,
    };
  }

  return {
    answer:
      analysis.red_flags[0] ??
      "The preview did not surface major launch-rule red flags. A full AI review should still confirm obligations, renewal terms, and dispute handling.",
    suggested_next_step: firstAction,
  };
}
